src/lem/README.md

```diff
@@ -0,0 +1,94 @@
+## Lurk Evaluation Model (LEM)
+
+The main purpose of LEM is to represent Lurk's step function *as data*, which can
+be interpreted or used to synthesize the circuit.
+
+### Pointer cycle
+
+The variables of a LEM are represented by *meta pointers*, which become bound to
+concrete pointers during interpretation. So a `MetaPtr` is a named reference to
+a `Ptr`.
+
+To synthesize the circuit, the `Ptr`s that come from interpretation are hashed
+and become `ZPtr`s, whose data can be allocated in the circuit to generate
+*allocated pointers* (`AllocatedPtr`).
+
+```mermaid
+graph TD;
+    M[MetaPtr] --> P[Ptr];
+    P --> Z[ZPtr];
+    Z --> A[AllocatedPtr];
+```
+
+### Check
+
+We statically check the following properties:
+
+* Static single assignments
+* Non-duplicated input labels
+* One return per LEM path
+* Assign first, use later
+* Ensure always return
+
+### LEM macro syntax
+
+A summary of LEM syntax:
+
+* `let`
+* `return`
+* `match_tag`
+* `match_symbol`
+* `hash`, `unhash`
+* `hide`, `open`
+
+# Static Analysis
+
+The first traversal of LEM is the static analysis phase, where we allocate hash slots already considering deduplication. 
+
+## STEP 1
+
+    * calculate hash slots for all virtual paths.
+
+# Interpretation
+
+The second traversal occurs during interpretation of LEM. Given as input an expression, an environment and a continuation, the concrete path is followed by the interpreter, computing the output. 
+
+## STEP 2
+
+    * calculate all preimages and images for all virtual paths.
```

Comment:
Can we not use the word `image` if we mean the thing we call 'hash' throughout the Lurk code base? We should try to converge on linguisitic/conceptual coherence. I would like this LEM work not to create shadow vocabulary/concepts any more than needed.

Reply:
"hash" or "digest" is more compatible with slots that are about poseidon hashing. But we want to support bit decompositions as well

Reply:
Also, what do you mean 'calculate' here? If virtual paths are those not followed by the interpretation, what is being calculated? Naively, my expectation is that we only 'calculate' the values associated with what you call 'concrete paths'.